README.md
# erdtree (erd)

[![Build status](https://github.com/solidiquis/erdtree/actions/workflows/ci.yml/badge.svg)](https://github.com/solidiquis/erdtree/actions)
[![Crates.io](https://img.shields.io/crates/v/erdtree.svg)](https://crates.io/crates/erdtree)
[![Packaging status](https://repology.org/badge/tiny-repos/erdtree.svg)](https://repology.org/project/erdtree/versions)
[![Crates.io](https://img.shields.io/crates/d/erdtree)](https://crates.io/crates/erdtree)

`erdtree` is a modern, cross-platform, multi-threaded, and general purpose filesystem and disk-usage utility that is aware of `.gitignore` and hidden file rules.
The following are some feature highlights:
* Reports disk usage using a variety of metrics: bytes (logical or physical), blocks (Unix-only), word-count, or line-count.
* Supports an `ls -l`-like view with information about owners, group, file permission, etc. (Unix-only).
* Respects hidden file and gitignore rules by default.
* Supports regular expressions and glob based searching by file-type.
* Comes with several layouts: a reverse tree output, a `tree`-like output, or a `du`-like output.
* Granular sorting capabilities.
* Supports icons.
* Colorized with `LS_COLORS`.

You can think of `erdtree` as a little bit of `du`, `tree`, `find`, `wc` and `ls`.

<p align="center">
  <img src="https://github.com/solidiquis/erdtree/blob/master/assets/top_showcase.png?raw=true" alt="failed to load picture" />
</p>

## Table of Contents

* [Usage](#usage)
* [Installation](#installation)
* [Documentation](#documentation)
  - [Configuration file](#configuration-file)
      - [Toml file](#toml-file)
      - [.erdtreerc](#erdtreerc)
  - [Hardlinks](#hardlinks)
  - [Symlinks](#symlinks)
  - [Disk usage](#disk-usage)
      - [Physical vs logical](#physical-vs-logical)
      - [Matching `du` output](#matching-du-output)
      - [Word and line count](#word-and-line-count)
  - [Layouts](#layouts)
  - [gitignore](#gitignore)
  - [Hidden files](#hidden-files)
  - [Icons](#icons)
  - [Maximum depth](#maximum-depth)
  - [Pruning empty directories](#pruning-empty-directories)
  - [Sorting](#sorting)
  - [Directories only](#directories-only)
  - [Long view](#long-view)
  - [Regular expressions and globbing](#regular-expressions-and-globbing)
  - [Truncating output](#truncating-output)
  - [Redirecting output and colorization](#redirecting-output-and-colorization)
  - [Parallelism](#parallelism)
  - [Completions](#completions)
  - [Same filesystem](#same-filesystem)
* [Comparisons against similar programs](#comparisons-against-similar-programs)
  - [exa](#exa)
  - [dua](#dua)
  - [dust](#dust)
  - [fd](#fd)
* [Rules for contributing](#rules-for-contributing)
* [Security policy](#security-policy)
* [Questions you might have](#questions-you-might-have)

## Usage

```
$ erd --help
erdtree (erd) is a cross-platform, multi-threaded, and general purpose filesystem and disk usage utility.

Usage: erd [OPTIONS] [DIR]

Arguments:
  [DIR]
          Directory to traverse; defaults to current working directory

Options:
  -c, --config <CONFIG>
          Use configuration of named table rather than the top-level table in .erdtree.toml

  -C, --color <COLOR>
          Mode of coloring output
          
          [default: auto]

          Possible values:
          - none:  Print plainly without ANSI escapes
          - auto:  Attempt to colorize output
          - force: Turn on colorization always

  -d, --disk-usage <DISK_USAGE>
          Print physical or logical file size
          
          [default: physical]

          Possible values:
          - logical:
            How many bytes does a file contain
          - physical:
            How many actual bytes on disk, taking into account blocks, sparse files, and compression
          - line:
            How many total lines a file contains
          - word:
            How many total words a file contains
          - block:
            How many blocks are allocated to store the file

  -f, --follow
          Follow symlinks

  -H, --human
          Print disk usage in human-readable format

  -i, --no-ignore
          Do not respect .gitignore files

  -I, --icons
          Display file icons

  -l, --long
          Show extended metadata and attributes

      --group
          Show file's groups

      --ino
          Show each file's ino

      --nlink
          Show the total number of hardlinks to the underlying inode

      --octal
          Show permissions in numeric octal format instead of symbolic

      --time <TIME>
          Which kind of timestamp to use; modified by default

          Possible values:
          - create: Time created (alias: ctime)
          - access: Time last accessed (alias: atime)
          - mod:    Time last modified (alias: mtime)

      --time-format <TIME_FORMAT>
          Which format to use for the timestamp; default by default

          Possible values:
          - iso:
            Timestamp formatted following the iso8601, with slight differences and the time-zone omitted
          - iso-strict:
            Timestamp formatted following the exact iso8601 specifications
          - short:
            Timestamp only shows date without time in YYYY-MM-DD format
          - default:
            Timestamp is shown in DD MMM HH:MM format

  -L, --level <NUM>
          Maximum depth to display

  -p, --pattern <PATTERN>
          Regular expression (or glob if '--glob' or '--iglob' is used) used to match files

      --glob
          Enables glob based searching

      --iglob
          Enables case-insensitive glob based searching

  -t, --file-type <FILE_TYPE>
          Restrict regex or glob search to a particular file-type

          Possible values:
          - file: A regular file
          - dir:  A directory
          - link: A symlink

  -P, --prune
          Remove empty directories from output

  -s, --sort <SORT>
          How to sort entries
          
          [default: size]

          Possible values:
          - name:    Sort entries by file name in lexicographical order
          - rname:   Sort entries by file name in reversed lexicographical order
          - size:    Sort entries by size smallest to largest, top to bottom
          - rsize:   Sort entries by size largest to smallest, bottom to top
          - access:  Sort entries by newer to older Accessing Date
          - raccess: Sort entries by older to newer Accessing Date
          - create:  Sort entries by newer to older Creation Date
          - rcreate: Sort entries by older to newer Creation Date
          - mod:     Sort entries by newer to older Alteration Date
          - rmod:    Sort entries by older to newer Alteration Date

      --dir-order <DIR_ORDER>
          Sort directories before or after all other file types
          
          [default: none]

          Possible values:
          - none:  Directories are ordered as if they were regular nodes
          - first: Sort directories above files
          - last:  Sort directories below files

  -T, --threads <THREADS>
          Number of threads to use
          
          [default: 10]

  -u, --unit <UNIT>
          Report disk usage in binary or SI units
          
          [default: bin]

          Possible values:
          - bin: Displays disk usage using binary prefixes
          - si:  Displays disk usage using SI prefixes

  -x, --one-file-system
          Prevent traversal into directories that are on different filesystems

  -y, --layout <LAYOUT>
          Which kind of layout to use when rendering the output
          
          [default: regular]

          Possible values:
          - regular:  Outputs the tree with the root node at the bottom of the output
          - inverted: Outputs the tree with the root node at the top of the output
          - flat:     Outputs a flat layout using paths rather than an ASCII tree
          - iflat:    Outputs an inverted flat layout with the root at the top of the output

  -., --hidden
          Show hidden files

      --no-git
          Disable traversal of .git directory when traversing hidden files

      --completions <COMPLETIONS>
          Print completions for a given shell to stdout
          
          [possible values: bash, elvish, fish, powershell, zsh]

      --dirs-only
          Only print directories

      --no-config
          Don't read configuration file

      --no-progress
          Hides the progress indicator

      --suppress-size
          Omit disk usage from output

      --truncate
          Truncate output to fit terminal emulator window

  -h, --help
          Print help (see a summary with '-h')

  -V, --version
          Print version
```

`-l, --long` and all of its arguments are currently not available on Windows, but support for a Windows variant is planned.

## Installation

### crates.io (non-Windows)

Make sure you have [Rust and its toolchain](https://www.rust-lang.org/tools/install) installed.

```
$ cargo install erdtree
```

### crates.io (Windows)

The Windows version relies on some experimental features in order to properly support hard-link detection. If you want to build from `crates.io` you'll first need to install the nightly toolchain before installing `erdtree`:

```
$ rustup toolchain install nightly-2023-06-11
```

Thereafter:

```
$ cargo +nightly-2023-06-11 install erdtree
```

### Homebrew-core

```
$ brew install erdtree
```

### Scoop

```
$ scoop install erdtree
```

### NetBSD

```
$ pkgin install erdtree
```


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
 
This project adheres to [Semantic Versioning](http://semver.org/).

## [3.1.2] - 2023-07-02

### Bug Fixes
- [Fix loading .erdtree.toml on Windows](https://github.com/solidiquis/erdtree/issues/217) [#218](https://github.com/solidiquis/erdtree/pull/218)
- [Fix race to print to stdout/stderr which made error messages not appear at times](https://github.com/solidiquis/erdtree/issues/219) [#220](https://github.com/solidiquis/erdtree/pull/220)

## [3.1.1] - 2023-07-01

[This patches](https://github.com/solidiquis/erdtree/pull/215) a deadlock that occurs when `--pattern` fails to make any matches and the progress indicator is enabled which causes `erdtree` to completely freeze.

## [3.1.0] - 2023-07-01

### [What's new](https://github.com/solidiquis/erdtree/pull/202)
- [Multiple configs with .erdtree.toml](https://github.com/solidiquis/erdtree/pull/201)
- [Increased performance with --suppress-size](https://github.com/solidiquis/erdtree/pull/203)
- [-x, --one-file-system](https://github.com/solidiquis/erdtree/pull/204)
- [Support for NO_COLOR](https://github.com/solidiquis/erdtree/pull/205)
- [Inverted flat layout](https://github.com/solidiquis/erdtree/pull/206)
- [Aliases: atime, ctime, and mtime](https://github.com/solidiquis/erdtree/pull/207)
- [ctrlc handler to restore cursor if progress indicator is enabled](https://github.com/solidiquis/erdtree/pull/210)


### [.erdtree.toml](https://github.com/solidiquis/erdtree/pull/201)

First and foremost:

**`.erdtreerc` is planned for deprecation by v3.3 so please migrate to `.erdtree.toml` by then.**

There is now support for multiple configs:

```
-c, --config <CONFIG>  Use configuration of named table rather than the top-level table in .erdtree.toml
```

So given the following example `.erdtree.toml`:

```toml
icons = true
human = true

# Compute file sizes like `du`
[du]
disk_usage = "block"
icons = true
layout = "flat"
no-ignore = true
no-git = true
hidden = true
level = 1

# Do as `ls -l`
[ls]
icons = true
human = true
level = 1
suppress-size = true
long = true
no-ignore = true
hidden = true

# How many lines of Rust are in this code base?
[rs]
disk-usage = "line"
level = 1
pattern = "\\.rs$"
```

```
$ erd

# equivalant to 

$ erd --human --icons
```

...

```
$ erd -c ls

# equivalent to

$ erd --icons --human --level 1 --suppress-size --long --no-ignore --hidden
```

etc.

For further information on how to get started with the new `.erdtree.toml` please refer to the [README.md](README.md#toml-file).

**NOTE**:
- Multiple configs are only supported via `.erdtree.toml`.
- If you have both a `.erdtree.toml` and `.erdtreerc` then the latter will take precedent. Attempts to use `--config` in said case will result in an error and a warning asking to migrate from `.erdtreerc` to `.erdtree.toml`.
- Again: **`.erdtreerc` will be deprecated in favor of `.erdtree.toml` by v3.3**


## [3.0.2] - 2023-06-12
- Added support for `.erdtree.toml`

### Bug fixes
- [Fixed segmentation fault that occurs with `--long` when file has an invalid user or group](https://github.com/solidiquis/erdtree/pull/193)
- [Updated deprecated Nerd symbols](https://github.com/solidiquis/erdtree/pull/192)
- [Change CJK and emoji block icons](https://github.com/solidiquis/erdtree/pull/194)

**Full Changelog**: https://github.com/solidiquis/erdtree/compare/v3.0.1...v3.0.2

## [3.0.1] - 2023-05-28

### Bug fixes
- Fixed issue where empty directories were automatically being pruned despite not opting into `--prune`.
    - https://github.com/solidiquis/erdtree/issues/185
- Fixed issue where searching via globs or regex doesn't work when user has a config file.

### Other changes:
- Added exit codes to `erdtree`.
- Progress indicator now acquires a lock to `stdout` reduce likelihood of printing the progress-text being a bottleneck.
- Some code-cleanup.

### Pull Requests
- https://github.com/solidiquis/erdtree/pull/183
- https://github.com/solidiquis/erdtree/pull/186
- https://github.com/solidiquis/erdtree/pull/187

### Contributors
- [KP64](https://github.com/KP64)

**Full Changelog**: https://github.com/solidiquis/erdtree/compare/v3.0.0...v3.0.1

## [3.0.0] - 2023-05-25

### Bug fixes
- [Completions now properly reference the new binary name](https://github.com/solidiquis/erdtree/commit/c4ff0dacb137f1007af27f692f1e5ee66bf47cb6), `erd`
    - https://github.com/solidiquis/erdtree/issues/141
    - https://github.com/solidiquis/erdtree/issues/143
    - https://github.com/solidiquis/erdtree/issues/150
- [Symlink target now shows full path rather than just file name](https://github.com/solidiquis/erdtree/pull/162)
    - https://github.com/solidiquis/erdtree/issues/148

src/icons.rs
use crate::hash;
use ansi_term::Style;
use ansi_term::{ANSIGenericString, Color};
use ignore::DirEntry;
use once_cell::sync::Lazy;
use std::{
    borrow::Cow,
    collections::HashMap,
    ffi::{OsStr, OsString},
    fs::FileType,
    path::Path,
};

/// Computes a plain, colorless icon with given parameters.
///
/// The precedent from highest to lowest in terms of which parameters determine the icon used
/// is as followed: file-type, file-extension, and then file-name. If an icon cannot be
/// computed the fall-back default icon is used.
///
/// If a directory entry is a link and the link target is provided, the link target will be
/// used to determine the icon.
pub fn compute(entry: &DirEntry, link_target: Option<&Path>) -> Cow<'static, str> {
    let icon = entry
        .file_type()
        .and_then(icon_from_file_type)
        .map(Cow::from);

    if let Some(i) = icon {
        return i;
    }

    let ext = match link_target {
        Some(target) if entry.path_is_symlink() => target.extension(),
        _ => entry.path().extension(),
    };

    let icon = ext.and_then(icon_from_ext).map(|(_, i)| Cow::from(i));

    if let Some(i) = icon {
        return i;
    }

    let icon = icon_from_file_name(entry.file_name()).map(Cow::from);

    if let Some(i) = icon {
        return i;
    }

    Cow::from(get_default_icon().1)
}

/// Computes a plain, colored icon with given parameters. See [compute] for more details.
pub fn compute_with_color(
    entry: &DirEntry,
    link_target: Option<&Path>,
    style: Option<Style>,
) -> Cow<'static, str> {
    let icon = entry
        .file_type()
        .and_then(icon_from_file_type)
        .map(Cow::from);

    let paint_icon = |icon| match style {
        Some(Style {
            foreground: Some(fg),
            ..
        }) => {
            let ansi_string: ANSIGenericString<str> = fg.bold().paint(icon);
            let styled_icon = ansi_string.to_string();
            Cow::from(styled_icon)
        }
        _ => icon,
    };

    if let Some(icon) = icon {
        return paint_icon(icon);
    }

    let ext = match link_target {
        Some(target) if entry.path_is_symlink() => target.extension(),
        _ => entry.path().extension(),
    };

    let icon = ext
        .and_then(icon_from_ext)
        .map(|attrs| Cow::from(col(attrs.0, attrs.1)));

    if let Some(i) = icon {
        return i;
    }

    let icon = icon_from_file_name(entry.file_name())
        .map(Cow::from)
        .map(paint_icon);

    if let Some(i) = icon {
        return i;
    }

    let (code, icon) = get_default_icon();
    Cow::from(col(code, icon))
}

/// Attempts to return an icon given a file extension along with its default color code 8-bit
/// value.
fn icon_from_ext(ext: &OsStr) -> Option<(u8, &'static str)> {
    EXT_ICON_MAP.get(ext).map(|(code, icon)| (*code, *icon))
}

/// Attempts to return an icon based on file type.
fn icon_from_file_type(ft: FileType) -> Option<&'static str> {
    if ft.is_dir() {
        return FILE_TYPE_ICON_MAP.get("dir").copied();
    } else if ft.is_symlink() {
        return FILE_TYPE_ICON_MAP.get("symlink").copied();
    }

    None
}

/// Attempts to get the icon associated with the special file kind.
fn icon_from_file_name(name: &OsStr) -> Option<&'static str> {
    FILE_NAME_ICON_MAP.get(name).copied()
}

/// Returns the default fallback icon.
fn get_default_icon<'a>() -> (u8, &'a str) {
    *DEFAULT_ICON
}

/// Convenience method to paint fixed colors.
fn col(num: u8, code: &str) -> String {
    Color::Fixed(num).paint(code).to_string()
}

/// Default fallback icon.
static DEFAULT_ICON: Lazy<(u8, &str)> = Lazy::new(|| (66, "\u{f15b}"));

/// Lazily evaluated static hash-map of special file-types and their corresponding styled icons.
/// These icons will take on the color properties of their associated file which is based on
/// `LS_COLORS`.
///
/// Dev icons sourced from [`exa`](https://github.com/ogham/exa/blob/master/src/output/icons.rs)
static FILE_TYPE_ICON_MAP: Lazy<HashMap<&str, &str>> = Lazy::new(|| {
    hash!(
        "dir"     => "\u{f413}", // 
        "symlink" => "\u{f482}"  // 
    )
});

/// Lazily evaluated static hash-map of special named and their corresponding icons. These icons
/// will take on the color properties of their associated file which is based on `LS_COLORS`.
///
/// Dev icons sourced from [`exa`](https://github.com/ogham/exa/blob/master/src/output/icons.rs)
static FILE_NAME_ICON_MAP: Lazy<HashMap<OsString, &str>> = Lazy::new(|| {
    hash!(
        OsString::from(".Trash")             => "\u{f1f8}", // 
        OsString::from(".atom")              => "\u{e764}", // 
        OsString::from(".bashprofile")       => "\u{e615}", // 
        OsString::from(".bashrc")            => "\u{f489}", // 
        OsString::from(".git")               => "\u{f1d3}", // 
        OsString::from(".gitattributes")     => "\u{f1d3}", // 
        OsString::from(".gitconfig")         => "\u{f1d3}", // 
        OsString::from(".github")            => "\u{f408}", // 
        OsString::from(".gitignore")         => "\u{f1d3}", // 
        OsString::from(".gitmodules")        => "\u{f1d3}", // 
        OsString::from(".rvm")               => "\u{e21e}", // 
        OsString::from(".vimrc")             => "\u{e62b}", // 
        OsString::from(".vscode")            => "\u{e70c}", // 
        OsString::from(".zshrc")             => "\u{f489}", // 
        OsString::from("Cargo.lock")         => "\u{e7a8}", // 
        OsString::from("bin")                => "\u{e5fc}", // 
        OsString::from("config")             => "\u{e5fc}", // 
        OsString::from("docker-compose.yml") => "\u{f308}", // 
        OsString::from("Dockerfile")         => "\u{f308}", // 
        OsString::from(".DS_Store")          => "\u{f179}", // 
        OsString::from("gitignore_global")   => "\u{f1d3}", // 
        OsString::from("go.mod")             => "\u{e626}", // 
        OsString::from("go.sum")             => "\u{e626}", // 
        OsString::from("gradle")             => "\u{e256}", // 
        OsString::from("gruntfile.coffee")   => "\u{e611}", // 
        OsString::from("gruntfile.js")       => "\u{e611}", // 
        OsString::from("gruntfile.ls")       => "\u{e611}", // 
        OsString::from("gulpfile.coffee")    => "\u{e610}", // 
        OsString::from("gulpfile.js")        => "\u{e610}", // 
        OsString::from("gulpfile.ls")        => "\u{e610}", // 
        OsString::from("hidden")             => "\u{f023}", // 
        OsString::from("include")            => "\u{e5fc}", // 
        OsString::from("lib")                => "\u{f121}", // 
        OsString::from("license")            => "\u{e60a}",   // 
        OsString::from("LICENSE")            => "\u{e60a}",   // 
        OsString::from("licence")            => "\u{e60a}",   // 
        OsString::from("LICENCE")            => "\u{e60a}",   // 
        OsString::from("localized")          => "\u{f179}", // 
        OsString::from("Makefile")           => "\u{f489}", // 
        OsString::from("node_modules")       => "\u{e718}", // 
        OsString::from("npmignore")          => "\u{e71e}", // 
        OsString::from("PKGBUILD")           => "\u{f303}", // 
        OsString::from("rubydoc")            => "\u{e73b}", // 
        OsString::from("yarn.lock")          => "\u{e718}"  // 
    )
});

/// Lazily evaluated static hash-map of various file extensions and their corresponding icons. The
/// key is the file extension while the associated value is a tuple containing the 8-bit color code
/// as well as the Unicode scalar value for the corresponding icon.
///
/// Dev icons and their color palettes sourced from [`nvim-web-devicons`](https://github.com/nvim-tree/nvim-web-devicons/blob/master/lua/nvim-web-devicons.lua).
static EXT_ICON_MAP: Lazy<HashMap<OsString, (u8, &str)>> = Lazy::new(|| {
    hash!(
        OsString::from("ai")            => (185, "\u{e7b4}"),   // 
        OsString::from("awk")           => (59, "\u{e795}"),    // 
        OsString::from("bash")          => (113, "\u{e795}"),   // 
        OsString::from("bat")           => (154, "\u{e615}"),   // 
        OsString::from("bmp")           => (140, "\u{e60d}"),   // 
        OsString::from("cbl")           => (25, "\u{2699}"),    // ⚙
        OsString::from("c++")           => (204, "\u{e61d}"),   // 
        OsString::from("c")             => (75, "\u{e61e}"),    // 
        OsString::from("cc")            => (204, "\u{e61d}"),   // 
        OsString::from("cfg")           => (231, "\u{e7a3}"),   // 
        OsString::from("cljc")          => (107, "\u{e768}"),   // 
        OsString::from("clj")           => (107, "\u{e768}"),   // 
        OsString::from("cljd")          => (67, "\u{e76a}"),    // 
        OsString::from("cljs")          => (67, "\u{e76a}"),    // 
        OsString::from("cmake")         => (66, "\u{e615}"),    // 
        OsString::from("cob")           => (25, "\u{2699}"),    // ⚙
        OsString::from("cobol")         => (25, "\u{2699}"),    // ⚙
        OsString::from("coffee")        => (185, "\u{e61b}"),   // 
        OsString::from("conf")          => (66, "\u{e615}"),    // 
        OsString::from("config.ru")     => (52, "\u{e791}"),    // 
        OsString::from("cp")            => (67, "\u{e61d}"),    // 
        OsString::from("cpp")           => (67, "\u{e61d}"),    // 
        OsString::from("cpy")           => (25, "\u{2699}"),    // ⚙
        OsString::from("cr")            => (16, "\u{e24f}"),    // 
        OsString::from("cs")            => (58, "\u{f81a}"),    // 
        OsString::from("csh")           => (59, "\u{e795}"),    // 
        OsString::from("cson")          => (185, "\u{e60b}"),   // 
        OsString::from("css")           => (39, "\u{e749}"),    // 
        OsString::from("csv")           => (113, "\u{f718}"),   // 
        OsString::from("cxx")           => (67, "\u{e61d}"),    // 
        OsString::from("dart")          => (25, "\u{e798}"),    // 
        OsString::from("db")            => (188, "\u{e706}"),   // 
        OsString::from("d")             => (64, "\u{e7af}"),    // 
        OsString::from("desktop")       => (60, "\u{f108}"),    // 
        OsString::from("diff")          => (59, "\u{e728}"),    // 
        OsString::from("doc")           => (25, "\u{f72b}"),    // 
        OsString::from("drl")           => (217, "\u{e28c}"),   // 
        OsString::from("dropbox")       => (27, "\u{e707}"),    // 
        OsString::from("dump")          => (188, "\u{e706}"),   // 
        OsString::from("edn")           => (67, "\u{e76a}"),    // 
        OsString::from("eex")           => (140, "\u{e62d}"),   // 
        OsString::from("ejs")           => (185, "\u{e60e}"),   // 
        OsString::from("elm")           => (67, "\u{e62c}"),    // 
        OsString::from("epp")           => (255, "\u{e631}"),   // 
        OsString::from("erb")           => (52, "\u{e60e}"),    // 
        OsString::from("erl")           => (132, "\u{e7b1}"),   // 
        OsString::from("ex")            => (140, "\u{e62d}"),   // 
        OsString::from("exs")           => (140, "\u{e62d}"),   // 
        OsString::from("f#")            => (67, "\u{e7a7}"),    // 
        OsString::from("fish")          => (59, "\u{e795}"),    // 
        OsString::from("fnl")           => (230, "\u{1f31c}"),  // 🌜
        OsString::from("fs")            => (67, "\u{e7a7}"),    // 
        OsString::from("fsi")           => (67, "\u{e7a7}"),    // 
        OsString::from("fsscript")      => (67, "\u{e7a7}"),    // 
        OsString::from("fsx")           => (67, "\u{e7a7}"),    // 
        OsString::from("GNUmakefile")   => (66, "\u{e779}"),    // 
        OsString::from("gd")            => (66, "\u{e615}"),    // 
        OsString::from("gemspec")       => (52, "\u{e791}"),    // 
        OsString::from("gif")           => (140, "\u{e60d}"),   // 
        OsString::from("git")           => (202, "\u{e702}"),   // 
        OsString::from("glb")           => (215, "\u{f1b2}"),   // 
        OsString::from("go")            => (67, "\u{e627}"),    // 
        OsString::from("godot")         => (66, "\u{e7a3}"),    // 
        OsString::from("gql")           => (199, "\u{f20e}"),   // 
        OsString::from("graphql")       => (199, "\u{f20e}"),   // 
        OsString::from("haml")          => (188, "\u{e60e}"),   // 
        OsString::from("hbs")           => (208, "\u{e60f}"),   // 
        OsString::from("h")             => (140, "\u{f0fd}"),   // 
        OsString::from("heex")          => (140, "\u{e62d}"),   // 
        OsString::from("hh")            => (140, "\u{f0fd}"),   // 
        OsString::from("hpp")           => (140, "\u{f0fd}"),   // 
        OsString::from("hrl")           => (132, "\u{e7b1}"),   // 
        OsString::from("hs")            => (140, "\u{e61f}"),   // 
        OsString::from("htm")           => (166, "\u{e60e}"),   // 
        OsString::from("html")          => (202, "\u{e736}"),   // 
        OsString::from("hxx")           => (140, "\u{f0fd}"),   // 
        OsString::from("ico")           => (185, "\u{e60d}"),   // 
        OsString::from("import")        => (231, "\u{f0c6}"),   // 
        OsString::from("ini")           => (66, "\u{e615}"),    // 
        OsString::from("java")          => (167, "\u{e738}"),   // 
        OsString::from("jl")            => (133, "\u{e624}"),   // 
        OsString::from("jpeg")          => (140, "\u{e60d}"),   // 
        OsString::from("jpg")           => (140, "\u{e60d}"),   // 
        OsString::from("js")            => (185, "\u{e60c}"),   // 
        OsString::from("json5")         => (185, "\u{fb25}"),   // ﬥ
        OsString::from("json")          => (185, "\u{e60b}"),   // 
        OsString::from("jsx")           => (67, "\u{e625}"),    // 
        OsString::from("ksh")           => (59, "\u{e795}"),    // 
        OsString::from("kt")            => (99, "\u{e634}"),    // 
        OsString::from("kts")           => (99, "\u{e634}"),    // 
        OsString::from("leex")          => (140, "\u{e62d}"),   // 
        OsString::from("less")          => (60, "\u{e614}"),    // 
        OsString::from("lhs")           => (140, "\u{e61f}"),   // 
        OsString::from("license")       => (185, "\u{e60a}"),   // 
        OsString::from("licence")       => (185, "\u{e60a}"),   // 
        OsString::from("lock")          => (250, "\u{f13e}"),   // 
        OsString::from("log")           => (255, "\u{f831}"),   // 
        OsString::from("lua")           => (74, "\u{e620}"),    // 
        OsString::from("luau")          => (74, "\u{e620}"),    // 
        OsString::from("makefile")      => (66, "\u{e779}"),    // 
        OsString::from("markdown")      => (67, "\u{e609}"),    // 
        OsString::from("Makefile")      => (66, "\u{e779}"),    // 
        OsString::from("material")      => (132, "\u{f7f4}"),   // 
        OsString::from("md")            => (255, "\u{f48a}"),   // 
        OsString::from("mdx")           => (67, "\u{f48a}"),    // 
        OsString::from("mint")          => (108, "\u{f829}"),   // 
        OsString::from("mjs")           => (221, "\u{e60c}"),   // 
        OsString::from("mk")            => (66, "\u{e779}"),    // 
        OsString::from("ml")            => (173, "\u{3bb}"),    // λ
        OsString::from("mli")           => (173, "\u{3bb}"),    // λ
        OsString::from("mo")            => (99, "\u{221e}"),    // ∞
        OsString::from("mustache")      => (173, "\u{e60f}"),   // 
        OsString::from("nim")           => (220, "\u{1f451}"),  // 👑
        OsString::from("nix")           => (110, "\u{f313}"),   // 
        OsString::from("opus")          => (208, "\u{f722}"),   // 
        OsString::from("otf")           => (231, "\u{f031}"),   // 
        OsString::from("pck")           => (66, "\u{f487}"),    // 
        OsString::from("pdf")           => (124, "\u{f724}"),   // 
        OsString::from("php")           => (140, "\u{e608}"),   // 
        OsString::from("pl")            => (67, "\u{e769}"),    // 
        OsString::from("pm")            => (67, "\u{e769}"),    // 
        OsString::from("png")           => (140, "\u{e60d}"),   // 
        OsString::from("pp")            => (255, "\u{e631}"),   // 
        OsString::from("ppt")           => (167, "\u{f726}"),   // 
        OsString::from("prisma")        => (255, "\u{5351}"),   // 卑
        OsString::from("pro")           => (179, "\u{e7a1}"),   // 
        OsString::from("ps1")           => (69, "\u{f0a0a}"),   // 󰨊
        OsString::from("psb")           => (67, "\u{e7b8}"),    // 
        OsString::from("psd1")          => (105, "\u{f0a0a}"),  // 󰨊
        OsString::from("psd")           => (67, "\u{e7b8}"),    // 
        OsString::from("psm1")          => (105, "\u{f0a0a}"),  // 󰨊
        OsString::from("pyc")           => (67, "\u{e606}"),    // 
        OsString::from("py")            => (61, "\u{e606}"),    // 
        OsString::from("pyd")           => (67, "\u{e606}"),    // 
        OsString::from("pyo")           => (67, "\u{e606}"),    // 
        OsString::from("query")         => (154, "\u{e21c}"),   // 
        OsString::from("rake")          => (52, "\u{e791}"),    // 
        OsString::from("rb")            => (52, "\u{e791}"),    // 
        OsString::from("r")             => (65, "\u{fcd2}"),    // ﳒ
        OsString::from("rlib")          => (180, "\u{e7a8}"),   // 
        OsString::from("rmd")           => (67, "\u{e609}"),    // 
        OsString::from("rproj")         => (65, "\u{9276}"),    // 鉶
        OsString::from("rs")            => (180, "\u{e7a8}"),   // 
        OsString::from("rss")           => (215, "\u{e619}"),   // 
        OsString::from("sass")          => (204, "\u{e603}"),   // 
        OsString::from("sbt")           => (167, "\u{e737}"),   // 
        OsString::from("scala")         => (167, "\u{e737}"),   // 
        OsString::from("scm")           => (16, "\u{fb26}"),    // ﬦ
        OsString::from("scss")          => (204, "\u{e603}"),   // 
        OsString::from("sh")            => (59, "\u{e795}"),    // 
        OsString::from("sig")           => (173, "\u{3bb}"),    // λ
        OsString::from("slim")          => (166, "\u{e60e}"),   // 
        OsString::from("sln")           => (98, "\u{e70c}"),    // 
        OsString::from("sml")           => (173, "\u{3bb}"),    // λ
        OsString::from("sol")           => (67, "\u{fcb9}"),    // ﲹ
        OsString::from("sql")           => (188, "\u{e706}"),   // 
        OsString::from("sqlite3")       => (188, "\u{e706}"),   // 
        OsString::from("sqlite")        => (188, "\u{e706}"),   // 
        OsString::from("styl")          => (107, "\u{e600}"),   // 
        OsString::from("sublime")       => (98, "\u{e7aa}"),    // 
        OsString::from("suo")           => (98, "\u{e70c}"),    // 
        OsString::from("sv")            => (29, "\u{f85a}"),    // 
        OsString::from("svelte")        => (202, "\u{f260}"),   // 
        OsString::from("svg")           => (215, "\u{fc1f}"),   // ﰟ
        OsString::from("svh")           => (29, "\u{f85a}"),    // 
        OsString::from("swift")         => (173, "\u{e755}"),   // 
        OsString::from("tbc")           => (67, "\u{fbd1}"),    // ﯑
        OsString::from("t")             => (67, "\u{e769}"),    // 
        OsString::from("tcl")           => (67, "\u{fbd1}"),    // ﯑
        OsString::from("terminal")      => (71, "\u{f489}"),    // 
        OsString::from("test.js")       => (173, "\u{e60c}"),   // 
        OsString::from("tex")           => (58, "\u{fb68}"),    // ﭨ
        OsString::from("tf")            => (57, "\u{e2a6}"),    // 
        OsString::from("tfvars")        => (57, "\u{f15b}"),    // 
        OsString::from("toml")          => (66, "\u{e615}"),    // 
        OsString::from("tres")          => (185, "\u{e706}"),   // 
        OsString::from("ts")            => (67, "\u{e628}"),    // 
        OsString::from("tscn")          => (140, "\u{f880}"),   // 
        OsString::from("tsx")           => (67, "\u{e7ba}"),    // 
        OsString::from("twig")          => (107, "\u{e61c}"),   // 
        OsString::from("txt")           => (113, "\u{f718}"),   // 
        OsString::from("vala")          => (5, "\u{e69e}"),     // 
        OsString::from("v")             => (29, "\u{f85a}"),    // 
        OsString::from("vh")            => (29, "\u{f85a}"),    // 
        OsString::from("vhd")           => (29, "\u{f85a}"),    // 
        OsString::from("vhdl")          => (29, "\u{f85a}"),    // 
        OsString::from("vim")           => (29, "\u{e62b}"),    // 
        OsString::from("vue")           => (107, "\u{fd42}"),   // ﵂
        OsString::from("wasm")          => (99, "\u{e6a1}"),    // 
        OsString::from("webmanifest")   => (221, "\u{e60b}"),   // 
        OsString::from("webpack")       => (67, "\u{fc29}"),    // ﰩ
        OsString::from("webp")          => (140, "\u{e60d}"),   // 
        OsString::from("xcplayground")  => (173, "\u{e755}"),   // 
        OsString::from("xls")           => (23, "\u{f71a}"),    // 
        OsString::from("xml")           => (173, "\u{8b39}"),   // 謹
        OsString::from("xul")           => (173, "\u{e745}"),   // 
        OsString::from("yaml")          => (66, "\u{e615}"),    // 
        OsString::from("yml")           => (66, "\u{e615}"),    // 
        OsString::from("zig")           => (208, "\u{f0e7}"),   // 
        OsString::from("zsh")           => (113, "\u{e795}")    // 
    )
});
